Std/Data/List/SplitOnList.lean

```diff
@@ -0,0 +1,72 @@
+/-
+Copyright (c) 2024 Bulhwi Cha, Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bulhwi Cha, Mario Carneiro
+-/
+import Std.Data.Nat.Lemmas
+import Std.Data.List.Lemmas
+
+namespace List
+
+/-- Returns `(l₁, l₂)` for the first split `l = l₁ ++ l₂` such that `P l₂` returns true. -/
+def splitOnceP (P : List α → Bool) (l : List α) : Option (List α × List α) :=
+  if P l then
+    some ([], l)
+  else
+    match l with
+    | [] => none
+    | a :: l => (splitOnceP P l).map fun (l, r) => (a :: l, r)
+
+theorem eq_append_of_splitOnceP (P : List α → Bool) (l : List α) :
+    ∀ l₁ l₂, splitOnceP P l = some (l₁, l₂) → l = l₁ ++ l₂ := by
+  induction l with
+  | nil => simp [splitOnceP]
+  | cons a l ih =>
+    simp [splitOnceP]
+    if h : P (a :: l) then
+      simp [h]
+    else
+      simp [h]
+      match e : splitOnceP P l with
+      | none => simp
+      | some (lf, rt) => simp; apply ih; assumption
+
+theorem P_of_splitOnceP (P : List α → Bool) (l : List α) :
+    ∀ l₁ l₂, splitOnceP P l = some (l₁, l₂) → P l₂ := by
+  induction l with
+  | nil => simp [splitOnceP]
+  | cons a l ih =>
+    simp [splitOnceP]
+    if h : P (a :: l) then
+      simp [h]
+    else
+      simp [h]
+      match e : splitOnceP P l with
+      | none => simp
+      | some (lf, rt) => simp; apply ih; assumption
+
+/--
+Split a list at every occurrence of a separator list. The separators are not in the result.
+```
+[1, 1, 2, 3, 2, 4, 4].splitOnList [2, 3] = [[1, 1], [2, 4, 4]]
+```
+-/
+def splitOnList [BEq α] (l sep : List α) : List (List α) :=
```

Comment:
Could I ask what the application of this is? It doesn't seem like a particularly common function to need!

Reply:
This function is derived from specification of `String.splitOn`, see #495

Reply:
Got it, thanks. That makes sense.

Reply:
We need it to prove `String.splitOn_of_valid`, which is on the to-do list in `Std.Data.String.Lemmas`.